[SPARK] Fix ClassLoader handling for OpenLineageExtensionProvider

ddebowczyk92 opened the pull request:
### Problem

Closes: #3407

### Solution

The solution works around class loader conflicts by transferring the `OpenLineageExtensionProvider` class between class loaders, albeit using an illegal reflective access. Here's how it works:

1. Serialization: The bytecode of the OpenLineageExtensionProvider class is retrieved from the source class loader and serialized into a ByteBuffer.
2. Reflective Access: Using reflection, the defineClass method of the target class loader is invoked to load the serialized bytecode into its context.
3. Usage: Once defined, the class becomes accessible in the target class loader, resolving the visibility issue.

> **Note:** All schema changes require discussion. Please [link the issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) for context.

- [ ] Your change modifies the [core](https://github.com/OpenLineage/OpenLineage/blob/main/spec/OpenLineage.json) OpenLineage model
- [ ] Your change modifies one or more OpenLineage [facets](https://github.com/OpenLineage/OpenLineage/tree/main/spec/facets)

If you're contributing a new integration, please specify the scope of the integration and how/where it has been tested (e.g., Apache Spark integration supports `S3` and `GCS` filesystem operations, tested with AWS EMR).

### Checklist

- [x] You've [signed-off](https://github.com/OpenLineage/OpenLineage/blob/main/why-the-dco.md) your work
- [x] Your pull request title follows our [guidelines](https://github.com/OpenLineage/OpenLineage/blob/main/CONTRIBUTING.md#creating-pull-requests)
- [ ] Your changes are accompanied by tests (_if relevant_)
- [ ] Your change contains a [small diff](https://kurtisnusbaum.medium.com/stacked-diffs-keeping-phabricator-diffs-small-d9964f4dcfa6) and is self-contained
- [ ] You've updated any relevant documentation (_if relevant_)
- [ ] Your comment includes a one-liner for the changelog about the specific purpose of the change (_not required for changes to tests, docs, or CI config_)
- [ ] You've versioned the core OpenLineage model or facets according to [SchemaVer](https://docs.snowplowanalytics.com/docs/pipeline-components-and-applications/iglu/common-architecture/schemaver) (_if relevant_)
- [x] You've added a [header](https://github.com/OpenLineage/OpenLineage/tree/main/.github/header_templates.md) to source files (_if relevant_)

----
SPDX-License-Identifier: Apache-2.0\
Copyright 2018-2025 contributors to the OpenLineage project

pawel-big-lebowski commented on integration/spark/shared/src/main/java/io/openlineage/spark/agent/lifecycle/SparkOpenLineageExtensionVisitorWrapper.java:
```diff
@@ -190,10 +199,28 @@ private Optional<ImmutablePair<Object, Method>> getMethod(
+    // machinery.
+    // As a result, there is no reliable way to bypass the entire mechanism, even if
+    // no OpenLineageExtensionProvider implementations are present.
+    ClassLoader currentThreadClassloader = Thread.currentThread().getContextClassLoader();
```
why is it needed to instantiate `currentThreadClassloader` here and pass to another method? 
It is not used in this context at all, isn't it?

ddebowczyk92 replied:
fixed

pawel-big-lebowski commented on integration/spark/shared/src/main/java/io/openlineage/spark/agent/lifecycle/SparkOpenLineageExtensionVisitorWrapper.java:
```diff
@@ -190,10 +199,28 @@ private Optional<ImmutablePair<Object, Method>> getMethod(
   }
 
   private static List<Object> init(String testExtensionProvider)
```
Isn't it that `init` method is created anytime OpenLineage context is created and each time `ByteBuffer` with class bytes is read?

ddebowczyk92 replied:
done

pawel-big-lebowski reviewed (commented):
My understanding is that this code contains machinery to support corner-case scenario for loading classes with different classloaders. I think it would be great to write a code in a way that clearly distinguishes flow for a happy path from a corner case (code comments, separate methods, etc.)

ddebowczyk92 commented:
> My understanding is that this code contains machinery to support corner-case scenario for loading classes with different classloaders. I think it would be great to write a code in a way that clearly distinguishes flow for a happy path from a corner case (code comments, separate methods, etc.)

@pawel-big-lebowski it’s not entirely true; loading OpenLineageExtensionProvider to other classloaders isn’t a corner-case scenario. We load it to every classloader that hasn’t loaded it yet.

I added some comments and addressed your suggestions
